## Assignment 4
#### Tiffany Tang
#### NetID: tt2405

In [1]:
data <- read.csv("cell_id.tab.txt", sep = "\t", header = TRUE,row.names=1, na.strings = "")

# Remove rows where all values are NA
data <- na.omit(data)
head(data
# doesn't work, look at code below

ERROR: Error in read.table(file = file, header = header, sep = sep, quote = quote, : duplicate 'row.names' are not allowed


In [2]:

# Read the file
data <- read.csv("cell_id.tab.txt",sep = "\t", header = TRUE) 
# header = TRUE means the first row of the file contains column names

head(data) #just to look

cell.line  canonical_name resource      Amelogenin CSF1PO D13S317 D16S539
1 A-253      A-253          ATCC:CRL-7902 X          12,13  9,12    12     
2 NCI-BL2195 NCI-BL2195     ATCC:CRL-5956 X,Y        11     11      10,14  
3 LCL-WEI    LCL-WEI        DSM:ACC-218   X,Y        10,11  11,12   13     
4 Hs 212.M   Hs 212.M       ATCC:CRL-7173 X,Y        11,12  12      11     
5 QGP-1      QGP-1          JCRB:0183     X          10,12  13      10,12  
6 HLC-1      HLC-1          RCB:0083      X          7,12   8       9,12   
  D18S51 D19S433 D21S11 D2S1338 D3S1358 D5S818 D7S820 D8S1179 FGA   Penta TH01
1 0      0       0      0       0       11     10,12  0       0     0     6,8 
2 0      0       0      0       0       11,12  10,11  0       0     0     9.3 
3 0      0       0      0       0       11,13  12     0       0     0     9.3 
4 0      0       0      0       0       13,15  8,11   0       0     0     6,9 
5 13     0       29     0       14,17   12     12     14      21,23 E     6,9 
6 0      0       0      0       0       9,10   11,13  0       0     0     7,9 
  TPOX vWA     
1 8,12 16      
2 8,9  17,18   
3 8,10 17      
4 11   16,18   
5 8,11 14,17,18
6 12   16,17

In [5]:
# Count how many rows have '293T' 
#It will be in the first column
num_293T <- sum(grepl("293T", data[[1]], ignore.case = TRUE)) #use grep to find 293T
# data[[1]] returns value in the first column that match '293T or 293t' directly
# grepl returns true or false
# sum will sum up the total of true
print(num_293T)

[1] 3


In [14]:
names_h <- data[grep("H", data[[2]], ignore.case = TRUE),]
names_h

cell.line       canonical_name resource      Amelogenin CSF1PO D13S317
4    Hs 212.M        Hs 212.M       ATCC:CRL-7173 X,Y        11,12  12     
6    HLC-1           HLC-1          RCB:0083      X          7,12   8      
8    HCC2911         HCC2911        GNE:586180    X          8,13   12,13  
9    Hs 523.Sk       Hs 523.Sk      ATCC:CRL-7314 X,Y        9,10   12,13  
10   HCC78           HCC78          GNE:586319    X,Y        11,12  8,12   
15   Cri du Chat     Cri du Chat    ATCC:CCL-90   X          10,12  11     
17   NCI-H747        NCI-H747       ATCC:CCL-252  X,Y        11,12  11,12  
19   LCLC-103H       LCLC-103H      GNE:586326    X,Y        10,12  11     
21   HCC1354         HCC1354        GNE:587050    X          8,12   11,12  
25   NCI-H810        NCI-H810       ATCC:CRL-5816 X,Y        10     12,13  
33   HE52            HE52           RCB:2268      X          12     8,12   
37   JHH-4           JHH-4          GNE:586216    X          11,12  12     
43   HS 764.Mu       Hs 764.Mu      ATCC:CRL-7494 X,Y        10,11  11     
45   HARA            HARA           JCRB:1080.0   X,Y        13     9      
47   Hs 618.T        Hs 618.T       ATCC:CRL-7380 X          12     12,13  
51   NCI-H1688       NCI-H1688      ATCC:CCL-257  X          12     8      
57   HCC1011         HCC1011        GNE:586408    X          12     10,12  
63   Hs 156.T        Hs 156.T       ATCC:CRL-7102 X,Y        11,12  12,14  
65   WSU-NHL         WSU-NHL        GNE:586984    X          11,13  8,10,11
71   NCI-H1437       NCI-H1437      GNE:129264    X          11,12  12     
81   HTST            HTST           RCB:0967      X          10,12  7,12   
83   hTERT BJ        hTERT BJ       GNE:133566    X,Y        10,12  8,9    
103  HAAE-2          HAAE-2         ATCC:CRL-2473 X,Y        10,11  11,12  
107  Hs 782.Sk       Hs 782.Sk      ATCC:CRL-7513 X          10     9,11   
109  NCI-H2030       NCI-H2030      GNE:131775    X          11     12     
117  NCI-H2171       NCI-H2171      GNE:130849    X,Y        10     11     
121  HE12            HE12           RCB:2233      X,Y        12,13  8      
129  HCC2157         HCC2157        GNE:587120    X          10     12     
131  H322T           H322T          GNE:131975    X          12     12     
135  NCI-H1666       NCI-H1666      GNE:130538    X          11     15     
⋮    ⋮               ⋮              ⋮             ⋮          ⋮      ⋮      
4989 JHOM-2B         JHOM-2B        RCB:1682      X          10,12  11     
4993 NCI-H1944       NCI-H1944      GNE:129693    X          13     11     
5019 Hs 696.Sk       Hs 696.Sk      ATCC:CRL-7431 X,Y        11,14  9      
5023 Hs 371.T        Hs 371.T       ATCC:CRL-7256 X          10     11     
5025 NCI-H820        NCI-H820       GNE:585819    X,Y        11,12  12     
5029 NCC16           PHK16-0b       JCRB:0142     X,Y        12,13  11     
5031 Hs 864.Sk       Hs 864.Sk      ATCC:CRL-7599 X          10,12  9,13   
5039 DFCI-LU011      HCC4006        ATCC:CRL-2883 X,Y        10     11,12  
5041 HE54            HE54           RCB:2270      X,Y        11,12  8      
5045 HCC1171         HCC1171        GNE:586213    X          9,10   11     
5047 HCC2157 BL      HCC2157 BL     ATCC:CRL-2341 X          10,11  12     
5049 HOP-18          HOP-18         GNE:130762    X,Y        11,12  8,12   
5059 HuH-7           HuH-7          RCB:1366      X          11     10     
5081 HCA-1           HCA-1          GNE:586924    X          12     12     
5083 HE42            HE42           RCB:2261      X,Y        10,12  9      
5085 HuP-T3          HUP-T3         GNE:584433    X,Y        10     9,11   
5087 HE21            HE21           RCB:2240      X          10,12  8,11   
5089 Hs 925.Sk       Hs 925.Sk      ATCC:CRL-7676 X,Y        10,12  11,13  
5097 NCI-H2198       NCI-H2198      GNE:586089    X,Y        11     11     
5099 NIH:OVCAR-3     NIH:OVCAR-3    GNE:131419    X          11,12  12     
5101 HSC-5           HSC-5          GNE:586959    X,Y

In [17]:
names_h[grep("X,Y", names_h[,4], ignore.case = TRUE),]

cell.line    canonical_name resource       Amelogenin CSF1PO D13S317 
4    Hs 212.M     Hs 212.M       ATCC:CRL-7173  X,Y        11,12  12      
9    Hs 523.Sk    Hs 523.Sk      ATCC:CRL-7314  X,Y        9,10   12,13   
10   HCC78        HCC78          GNE:586319     X,Y        11,12  8,12    
17   NCI-H747     NCI-H747       ATCC:CCL-252   X,Y        11,12  11,12   
19   LCLC-103H    LCLC-103H      GNE:586326     X,Y        10,12  11      
25   NCI-H810     NCI-H810       ATCC:CRL-5816  X,Y        10     12,13   
43   HS 764.Mu    Hs 764.Mu      ATCC:CRL-7494  X,Y        10,11  11      
45   HARA         HARA           JCRB:1080.0    X,Y        13     9       
63   Hs 156.T     Hs 156.T       ATCC:CRL-7102  X,Y        11,12  12,14   
83   hTERT BJ     hTERT BJ       GNE:133566     X,Y        10,12  8,9     
103  HAAE-2       HAAE-2         ATCC:CRL-2473  X,Y        10,11  11,12   
117  NCI-H2171    NCI-H2171      GNE:130849     X,Y        10     11      
121  HE12         HE12           RCB:2233       X,Y        12,13  8       
151  Hs 697.Sp    Hs 697.Sp      ATCC:CRL-7433  X,Y        11     9,11    
159  HE33         HE33           RCB:2250       X,Y        10,12  11,12   
187  Hs 48.Fs     Hs 48.Fs       ATCC:CRL-7027  X,Y        10,12  12      
199  Hs 483.T     Hs 483.T       ATCC:CRL-7814  X,Y        10,12  12      
239  HE22         HE22           RCB:2241       X,Y        12     8,9     
277  NCI-H1672    NCI-H1672      GNE:586471     X,Y        10,11  11      
283  HFL-I        HFL-I          RCB:0521       X,Y        12,13  8,9     
287  Hs 890.Sk    Hs 890.Sk      ATCC:CRL-7627  X,Y        11,13  8,9     
301  THLE-2       THLE-2         ATCC:CRL-10149 X,Y        11,13  8,12    
317  Hs 683       Hs 683         GNE:132769     X,Y        9,13   8,12    
327  HE31         HE31           RCB:2248       X,Y        10,12  12      
377  NCI-BL1648   NCI-H1648      ATCC:CRL-5954  X,Y        10,12  12      
417  MHH-CALL-4   MHH-CALL-4     DSM:ACC-337    X,Y        11,12  9,10    
461  Hs 127.T     Hs 127.T       ATCC:CRL-7081  X,Y        11     12,13   
463  NCI-H1703    NCI-H1703      GNE:130536     X,Y        12     10      
487  HEL 299      HEL 299        ATCC:CCL-137   X,Y        7,10   11,12   
491  HE28         HE28           RCB:2245       X,Y        10,12  9,12    
⋮    ⋮            ⋮              ⋮              ⋮          ⋮      ⋮       
4767 HUVS-112D    HUVS-112D      ATCC:CRL-2481  X,Y        8,12   8,12    
4777 Hs 738.Lu    Hs 738.Lu      ATCC:CRL-7868  X,Y        10,11  8,11    
4779 NCI-H446     NCI-H446       GNE:585322     X,Y        13,14  8       
4781 HFb16d       HFb16d         JCRB:1008      X,Y        10,12  11      
4801 HuT 102      HuT 102        ATCC:TIB-162   X,Y        8,11   11,13   
4815 HCC345       HCC345         GNE:587152     X,Y        11,12  11,12   
4835 HCC630       HCC630         GNE:586516     X,Y        11     11,12,13
4847 Hs 168.Fs    Hs 168.Fs      ATCC:CRL-7122  X,Y        10,12  9,12    
4861 HUC-Fm2      HUC-Fm2        RCB:0437       X,Y        10,12  9,10    
4931 Hs 72        Hs 72.Fs       ATCC:CRL-7049  X,Y        11,13  9,10    
4935 Hs 728.Sk/Mu Hs 728.Sk/Mu   ATCC:CRL-7462  X,Y        10,11  11,12   
4937 Hs 80        Hs 80.Fs       ATCC:CRL-7058  X,Y        10,12  11,12   
4949 TUHR10TKB    TUHR10TKB      RCB:1275       X,Y        10,12  14      
4961 Hs 202.Sp    Hs 202.Sp      ATCC:CRL-7162  X,Y        11,12  11,12   
4969 Hs 674.Sk    Hs 674.Sk      ATCC:CRL-7397  X,Y        9,11   10      
5019 Hs 696.Sk    Hs 696.Sk      ATCC:CRL-7431  X,Y        11,14  9       
5025 NCI-H820     NCI-H820       GNE:585819     X,Y        11,12  12      
5029 NCC16        PHK16-0b       JCRB:0142      X,Y        12,13  11      
5039 DFCI-LU011   HCC4006        ATCC:CRL-2883  X,Y        10     11,12   
5041 HE54         HE54           RCB:2270       X,Y        11,12  8       
5049 HOP-18       HOP-18         GNE:130762     X,Y        11,12  8,12    
5083 HE42         HE42        

In [10]:
rows_293<- ("293T", data[[1]], ignore.case = TRUE)

data[grep("293T", data[[1]], ignore.case = TRUE),] # can put in with grep into the row


[1] 1099 3173 5017

cell.line canonical_name resource       Amelogenin CSF1PO D13S317 D16S539
1099 293T GNE  293            GNE:129641     X          11,12  12      9,13   
3173 293T/17   293            ATCC:CRL-11268 X          11,12  12,14   9,13   
5017 293T      293            DSM:ACC-635    X          11,12  12,14   9,13   
     D18S51 D19S433 D21S11  D2S1338 D3S1358 D5S818 D7S820 D8S1179     FGA  
1099 17,19  0       28,30.2 0       15,17   8,9    11,12  11,12,14,15 20,23
3173 0      0       0       0       0       8,9    11     0           0    
5017 0      0       0       0       0       8,9    11     0           0    
     Penta TH01  TPOX vWA     
1099 E     7,9.3 11   16,19   
3173 0     7,9.3 11   16,18,19
5017 0     7,9.3 11   16,19

There are 3 rows that have 293T in their name

In [6]:
# Get the name of the marker in column 7
marker_name <- colnames(data)[7]
# colnames(data) retrieves the column names, [7] indexes the 7th item from the list.
print(marker_name)

[1] "D16S539"


The marker in column 7 is `D16S539`

In [7]:
# Entry in row 2, column 7
entry_row2_col7 <- data[2, 7]
# goes into data, row 2 column 7 and retrieves what is in that cell

print(entry_row2_col7)

[1] "10,14"


The entry is 10, 14

In [14]:
names(sort(table(data$D13S317), decreasing = TRUE))
#create a table of all values in the column D13S317 of 'data'.
#sort based on frequency, with the most frequent first- decreasing = TRUE
# names extracts the names of the sorted frequency table. 

[1] ""              "11"            "12"            "11,12"        
 [5] "8,11"          "8"             "8,12"          "11,13"        
 [9] "9,12"          "12,13"         "9,11"          "9"            
[13] "10,11"         "10,12"         "8,9"           "10"           
[17] "13"            "12,14"         "8,10"          "8,13"         
[21] "11,14"         "14"            "9,13"          "10,13"        
[25] "9,10"          "13,14"         "9,14"          "8,14"         
[29] "10,14"         "12,13.3"       "11,12,13"      "8,10,11"      
[33] "8,11,12"       "13.3"          "15"            "0"            
[37] "13.3,13.3"     "7,12"          "8,13,14"       "9,12,13"      
[41] "11,12,15"      "12,13,14"      "12,15"         "13,14,15"     
[45] "8,9,11"        "8,9,12"        "9,10,11"       "10,11,12"     
[49] "10,12,13"      "10,12,13.3"    "10,14,13"      "11,11"        
[53] "11,11.3,12"    "11,12,13,14"   "11,12,14"      "11,13,14"     
[57] "11,14,10"      "11,15"         "11,16"         "12,12"        
[61] "12,13,14,16"   "13,13"         "13.3,14.3"     "6"            
[65] "7"             "7,11"          "7,13"          "7.2,11"       
[69] "8,10,12"       "8,11,13,14"    "8,11,14"       "8,12,13"      
[73] "8,13,12"       "9,10,11,12"    "9,10,11,12,13" "9,10,12"      
[77] "9,10,13"

Most common data entry is: 11 and 12 at 322 entries each

In [15]:
sort(table(data$D13S317)) #to see the data


     10,11,12      10,12,13    10,12,13.3      10,14,13         11,11 
            1             1             1             1             1 
   11,11.3,12   11,12,13,14      11,12,14      11,13,14      11,14,10 
            1             1             1             1             1 
        11,15         11,16         12,12   12,13,14,16         13,13 
            1             1             1             1             1 
    13.3,14.3             6             7          7,11          7,13 
            1             1             1             1             1 
       7.2,11       8,10,12    8,11,13,14       8,11,14       8,12,13 
            1             1             1             1             1 
      8,13,12    9,10,11,12 9,10,11,12,13       9,10,12       9,10,13 
            1             1             1             1             1 
     11,12,15      12,13,14         12,15      13,14,15        8,9,11 
            2             2             2             2             2 
     

### Part II

In [16]:
#import the weight height data set
wh <- read.csv("weight-height.csv", header = TRUE) #here again, header = TRUE, first row is colnames

1) Randomly select 30 males and 30 females, 

   - a) Estimate the mean and its variance using jackknife,
       for the 30 males, 30 females and all 60 together


   - b) Estimate the mean and variance using bootstrap
       for the 30 males, 30 females and all 60 together


   - c)  Do the estimates change between 1000 and 10,000 samplings in bootstrap?


   - d) Are the numbers consistent between jackknife and bootstrap  ?

In [17]:
#sample 30 males and 30 females
# first I will split the wh by males and females

wh_males <- wh[wh$Gender == "Male", ]#this creates a dataframe of just males
wh_females <- wh[wh$Gender == "Female", ]#this creates a dataframe of just males

# Sample 30 males from the data set
# sample from 1 through all the rows of wh_males, select 30 without replacement
sample_males <- wh_males[sample(1:nrow(wh_males), 30, replace=FALSE), ]

# Sample 30 females from the data set
sample_females <- wh_females[sample(1:nrow(wh_females), 30, replace=FALSE), ]


In [18]:
head(sample_males)#Just to take a look

Gender Height   Weight  
2617 Male   71.45649 179.4432
2021 Male   67.52710 167.2096
3649 Male   66.46922 185.6926
38   Male   67.19352 183.8110
4615 Male   67.00858 168.9268
4677 Male   67.90758 173.8981

In [19]:
library("bootstrap") #import library

#for jackknife function, need to compute mean
# So susbsitute the theta for the mean function 

#create a mean function, where my arguments are data and indicies
# it computes the mean of the subset of data
mean_function <- function(data, indicies){
    return(mean(data[indicies]))
}

#put in the data and the function into the jackknife function
#Perform jackknife for males
# jackknife function is applied to columns in the sample_males data set, 
# using the mean_function as the function to compute the jackknife estimate
jackknife_males_height <- jackknife(sample_males$Height, mean_function)
jackknife_males_weight <- jackknife(sample_males$Weight, mean_function)


print("Jackknife estimate for the males mean, height then weight:")
print(mean(jackknife_males_height[[3]]))#this will give the estimate of theta, 3rd element is theta
print(mean(jackknife_males_weight[[3]]))
cat("\n")

# Perform jackknife for females
jackknife_females_height <- jackknife(sample_females$Height, mean_function)
jackknife_females_weight <- jackknife(sample_females$Weight, mean_function)
# Print out in the same manner
print("Jackknife estimate for the females mean, height then weight:")
print(mean(jackknife_females_height[[3]])) 
print(mean(jackknife_females_weight[[3]]))
cat("\n")

# Combine male and female data for the combined group
sample_combined <- rbind(sample_males, sample_females)

# Perform jackknife for combined group
jackknife_combined_height <- jackknife(sample_combined$Height, mean_function)
jackknife_combined_weight <- jackknife(sample_combined$Weight, mean_function)
# Print out in the same manner 
print("Jackknife estimate for the combined group mean, height then weight:")
print(mean(jackknife_combined_height[[3]]))
print(mean(jackknife_combined_weight[[3]])) # This gives the jackknife mean estimate

[1] "Jackknife estimate for the males mean, height then weight:"
[1] 69.53268
[1] 189.4182

[1] "Jackknife estimate for the females mean, height then weight:"
[1] 64.02559
[1] 138.2602

[1] "Jackknife estimate for the combined group mean, height then weight:"
[1] 66.77913
[1] 163.8392


In [20]:
# I am going to repeat the same thing, but instead of mean
# I will replace theta with a variance function. 

# Variance
variance_function <- function(data, indices) {
  return(var(data[indices]))  # Calculate variance for the selected indices
}

# Perform jackknife for males
# same as above, except perform jackknife using variance as the function
jackknife_males_height <- jackknife(sample_males$Height, variance_function)
jackknife_males_weight <- jackknife(sample_males$Weight, variance_function)

# Jackknife estimate of variance for males
print("Jackknife estimate for the males variance, height then weight variance:")
print(mean(jackknife_males_height[[3]]))  # This gives the jackknife estimate of variance for height
print(mean(jackknife_males_weight[[3]]))  # This gives the jackknife estimate of variance for weight
cat("\n")

# Perform jackknife for females
jackknife_females_height <- jackknife(sample_females$Height, variance_function)
jackknife_females_weight <- jackknife(sample_females$Weight, variance_function)

# Jackknife estimate of variance for females
print("Jackknife estimate for the females variance, height then weight variance:")
print(mean(jackknife_females_height[[3]]))  # Jackknife estimate for height variance
print(mean(jackknife_females_weight[[3]]))  # Jackknife estimate for weight variance
cat("\n")

# Combine male and female data for the combined group
sample_combined <- rbind(sample_males, sample_females)

# Perform jackknife for combined group
jackknife_combined_height <- jackknife(sample_combined$Height, variance_function)
jackknife_combined_weight <- jackknife(sample_combined$Weight, variance_function)

# Jackknife estimate of variance for combined group
print("Jackknife estimate for the combined group variance, height then weight variance:")
print(mean(jackknife_combined_height[[3]]))  # Jackknife estimate for combined height variance
print(mean(jackknife_combined_weight[[3]]))  # Jackknife estimate for combined weight variance

[1] "Jackknife estimate for the males variance, height then weight variance:"
[1] 8.018708
[1] 358.1387

[1] "Jackknife estimate for the females variance, height then weight variance:"
[1] 6.797423
[1] 337.9423

[1] "Jackknife estimate for the combined group variance, height then weight variance:"
[1] 14.99301
[1] 1007.516


In [21]:
# Estimate the mean and variance with bootstrap
# Load the bootstrap package
library(bootstrap)

# Perform bootstrap for males' height and weight
# just use the mean function as function calculate bootstrap
bootstrap_males_height <- bootstrap(sample_males$Height, 100, mean)  # Bootstrap for height
bootstrap_males_weight <- bootstrap(sample_males$Weight, 100, mean)  # Bootstrap for weight

# Print the bootstrap estimates for males
print("Bootstrap estimate for the males, height then weight:")
print(mean(bootstrap_males_height$thetastar))  # Mean of bootstrap estimates for height
print(mean(bootstrap_males_weight$thetastar))  # Mean of bootstrap estimates for weight
cat("\n")

# Perform bootstrap for remales' height and weight
bootstrap_females_height <- bootstrap(sample_females$Height, 100, mean)  # Bootstrap for height
bootstrap_females_weight <- bootstrap(sample_females$Weight, 100, mean)  # Bootstrap for weight

# Print the bootstrap estimates for females
print("Bootstrap estimate for the females, height then weight:")
print(mean(bootstrap_females_height$thetastar))  # Mean of bootstrap estimates for height
print(mean(bootstrap_females_weight$thetastar))  # Mean of bootstrap estimates for weight
cat("\n")

# Combine male and female data for the combined group
sample_combined <- rbind(sample_males, sample_females)
# Perform bootstrap for remales' height and weight
bootstrap_combined_height <- bootstrap(sample_combined$Height, 100, mean)  # Bootstrap for height
bootstrap_combined_weight <- bootstrap(sample_combined$Weight, 100, mean)  # Bootstrap for weight

# Print the bootstrap estimates for females
print("Bootstrap estimate for the combined groupes, height then weight:")
print(mean(bootstrap_combined_height$thetastar))  # Mean of bootstrap estimates for height
print(mean(bootstrap_combined_weight$thetastar))  # Mean of bootstrap estimates for weight
cat("\n")

[1] "Bootstrap estimate for the males, height then weight:"
[1] 69.56101
[1] 189.6205

[1] "Bootstrap estimate for the females, height then weight:"
[1] 64.00579
[1] 137.9125

[1] "Bootstrap estimate for the combined groupes, height then weight:"
[1] 66.71857
[1] 164.4282



In [22]:
#repeat for variance 

# Perform bootstrap for males' height and weight
# use variance as function to calculate bootstrap
bootstrap_males_height <- bootstrap(sample_males$Height, 100, var)  # Bootstrap for height
bootstrap_males_weight <- bootstrap(sample_males$Weight, 100, var)  # Bootstrap for weight

# Print the bootstrap estimates for males
print("Bootstrap estimate for the males, height then weight:")
print(mean(bootstrap_males_height$thetastar))  # Mean of bootstrap estimates for height
print(mean(bootstrap_males_weight$thetastar))  # Mean of bootstrap estimates for weight
cat("\n")

# Perform bootstrap for remales' height and weight
bootstrap_females_height <- bootstrap(sample_females$Height, 100, var)  # Bootstrap for height
bootstrap_females_weight <- bootstrap(sample_females$Weight, 100, var)  # Bootstrap for weight

# Print the bootstrap estimates for females
print("Bootstrap estimate for the females, height then weight:")
print(mean(bootstrap_females_height$thetastar))  # Mean of bootstrap estimates for height
print(mean(bootstrap_females_weight$thetastar))  # Mean of bootstrap estimates for weight
cat("\n")

# Combine male and female data for the combined group
sample_combined <- rbind(sample_males, sample_females)
# Perform bootstrap for remales' height and weight
bootstrap_combined_height <- bootstrap(sample_combined$Height, 100, var)  # Bootstrap for height
bootstrap_combined_weight <- bootstrap(sample_combined$Weight, 100, var)  # Bootstrap for weight

# Print the bootstrap estimates for females
print("Bootstrap estimate for the combined groupes, height then weight:")
print(mean(bootstrap_combined_height$thetastar))  # Mean of bootstrap estimates for height
print(mean(bootstrap_combined_weight$thetastar))  # Mean of bootstrap estimates for weight
cat("\n")

[1] "Bootstrap estimate for the males, height then weight:"
[1] 7.747354
[1] 360

[1] "Bootstrap estimate for the females, height then weight:"
[1] 6.452509
[1] 320.5531

[1] "Bootstrap estimate for the combined groupes, height then weight:"
[1] 14.949
[1] 1003.64



In [23]:
 # Do the estimates change between 1000 and 10,000 samplings in bootstrap?

#lets test for males height
test1 <- bootstrap(sample_males$Height, 1000, var)

test2 <- bootstrap(sample_males$Height, 10000, var)
print(mean(test1$thetastar))
print(mean(test2$thetastar))


[1] 7.757983
[1] 7.736265


There does not seem to be a massive difference in the estimates- only around .05 difference between the two values after running the above code a few times. 

In [24]:
# Can test with the means too
# Lets test for males height
test1 <- bootstrap(sample_males$Height, 1000, mean)

test2 <- bootstrap(sample_males$Height, 10000, mean)
print(mean(test1$thetastar))
print(mean(test2$thetastar))
# Getting similar results

[1] 69.55636
[1] 69.53261


In the above code it looks like the same, there is only a difference around 0.02 between the two estimates.

**Are the numbers consistent between jackknife and bootstrap**

Taking a look at our above results: 

- Means:
  
    `"Jackknife esitmate for the males, height then weight:" 69.27688, 189.674`
  
    `"Jackknife esitmate for the females, height then weight:" 63.0887, 136.7103`
  
    `"Jackknife esitmate for the combined group, height then weight:" 66.18279, 163.1921`

    `"Bootstrap estimate for the males, height then weight:" 69.28331, 189.3336`
  
    `"Bootstrap estimate for the females, height then weight:" 63.10928, 136.89`
  
    `"Bootstrap estimate for the combined groupes, height then weight:" 66.27431, 163.1933`

  Both estimates for the means are very consistent with each other, and similarly, the variance estimates show consistency as well. However, there is greater variability in the weight variance estimates, likely due to the wider spread of the variance data compared to the mean data.

**Using the sample function in R, write a bootstrap routine that outputs the bootstrap mean. Explain the logic by annotating the code.**

In [26]:
# Bootstrap mean function
# Bootstrap resamples the data with replacement
# Need to pull a sample the same size as the original dataset, 

# Define the function, inputs will be data, and # of times you want to resample
bootstrap_mean <- function(data, num_resamples){
    means_v <- numeric(num_resamples) #initializes the vector of all your means of your resamples

    # Resampling process
    for (i in 1: num_resamples){ # Will resample num_resamples times
        # Resample the number of times the length of the dataset
        resample <- sample( data, size = length(data), replace = TRUE) # With replacement!!!
        #boot strap samples with replacement!!

        # Since we want bootstrap with mean
        # calculate the mean of the resampled data
        means_v[i] <- mean(resample) # Add the mean into the means vector at the correct index
    }
    return( mean(means_v)) #return the mean of the means
}

In [27]:
# Let's test with the sample sets from the previous question
bootstrap_mean(sample_males$Height, 1000)
#output looks good!

[1] 69.52396

**Sample without replacement people's heights and weights
   from our weight-height data. Demonstrate the correlation between
   height and weight in this sample**

In [30]:
# cor(wh[,2],wh[,3],method=c("spearman"))gives us 0.9257077 for the full set

# Sample without replacement, we will sample 100 people randomly from the set
sample_random <- wh[sample(1:nrow(wh), 100, replace=FALSE), ]

# Perform the Spearman-rank correlation
spearman_sample <- cor(wh[,2], wh[,3], method =c("spearman")) 
# do spearman correlation on the 2nd and third columns
print(spearman_sample)

[1] 0.9257077


The correlation for our random sample is 0.9257, which is nearly the same as the correlation for the entire weight height data set. This is a very high correlation which suggests that an increase in height will means an increase in weight. Since the correlation of the sample is very similar to the correlation of the whole, it suggests that the sample is representative of the full data set. 